File: gfg/2024_8_august/c++/19_kth_smallest.cpp
// TIME COMPEXITY
// O(nlogk)

// SPACE COMPEXITY
// O(k)

// CODE

class Solution {
  public:
    // arr : given array
    // k : find kth smallest element and return using this function
    int kthSmallest(vector<int> &arr, int k) {
        priority_queue<int> maxHeap;
        for(int i = 0; i < arr.size(); i++){
            maxHeap.push(arr[i]);
            if(maxHeap.size() > k){
                maxHeap.pop();
            }
        }
        return maxHeap.top();
    }
};

File: gfg/2024_8_august/python/19_kth_smallest.py
# CODE

from heapq import heappush, heappop

class Solution:
    def kthSmallest(self, arr, k):
        n, i = len(arr), 0
        pq = []
        while i < k:
            heappush(pq, -arr[i])
            i += 1
        while i < n:
            if arr[i] < -pq[0]:
                heappop(pq)
                heappush(pq, -arr[i])
            i += 1
        return -heappop(pq)